### DISCUSSION TITLE
Expectation about activities!

### DISCUSSION THREAD

**Discussion by alibey-10** [statelyai/xstate#1652] [Q&A] [answered]
I tried to do "activities" to update the context with `assign` and `setInterval`, it works when logging, but it did not reflect the change of the context **in the react component**, so I worked around it and I used actions and sent events with `setInterval`.  the problem solved but still wondering why it did not work with "activities", was my expectation incorrect, or should I have debugged my code better?

**Comment by davidkpiano** [comment, marked as answer]
Can you share code?

If you want some external entity (an "activity" or whatever you want to call it) to signal to the machine that it should update context on an interval, you can do it like this:

```js
const machine = createMachine({
  // ...
  invoke: {
    src: () => sendBack => {
      const i = setInterval(() => {
        sendBack('UPDATE');
      }, 1000);
      
      return () => clearInterval(i);
    }
  },
  on: {
    UPDATE: { actions: assign({ count: ctx => ctx.count + 1 }) }
  }
});
```

**Reply by mjangir** [reply]
> Can you share code?
> 
> If you want some external entity (an "activity" or whatever you want to call it) to signal to the machine that it should update context on an interval, you can do it like this:
> 
> ```js
> const machine = createMachine({
>   // ...
>   invoke: {
>     src: () => sendBack => {
>       const i = setInterval(() => {
>         sendBack('UPDATE');
>       }, 1000);
>       
>       return () => clearInterval(i);
>     }
>   },
>   on: {
>     UPDATE: { actions: assign({ count: ctx => ctx.count + 1 }) }
>   }
> });
> ```

However, updating the context from an invoked callback is having some issues if the interval is 100 ms or like that. I'm unable to transition to another state if the setInterval is set to too short internal around 100-200ms.

**Comment by alibey-10** [comment]
here is my code,  
machine state 
```javascript
const machine = createMachine({
// ...
{
activities: {
  UPDATE: () => {
    const i = setInterval(() => {
      assign({
        value: (c) => c.value + 1
      });
    }, 1000);
    return () => clearInterval(i);
  },
})
```
React part
```javascript

const [state] =useMachine(machine)
const {value}= state.context
const [localstate, setLocalstate]= useState(value)

useEffect(()=>setLocalstate() {
const i = setInterval(()=>setLocalstate(value), 1000)
// cleanup...
}, [value])
// ...
```

**Reply by Andarist** [reply]
What you have posted here differs quite a bit from what @davidkpiano has proposed. Please try to use the proposed solution as it should work. You can't just call `assign` at a random time (within the internal) and expect something to happen. `assign` returns a description object that has to be given to the Interpreter (which happens automatically, but it has to be put at correct locations in your schema) so it can actually execute what has been "described".

**Reply by davidkpiano** [reply]
Right: `assign(...)` is a _pure function_; it doesn't do anything imperatively.

**Reply by alibey-10** [reply]
@Andarist yes, that was my code before @davidkpiano proposed his solution, just shared it upon his request. Thanks for the explanation.
